File: ruby/05-zombie-code/test/campaign_sender_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../src/campaign_sender'

class CampaignSenderTest < Minitest::Test
  def setup
    @sender = CampaignSender.new
  end

  def test_counts_eu_customers_with_opt_in
    result = @sender.send_campaign([{ active: true, region: 'EU', gdpr_opt_in: true }], 'Hi')
    assert_equal 1, result[:sent]
  end

  def test_skips_eu_customers_without_opt_in
    result = @sender.send_campaign([{ active: true, region: 'EU', gdpr_opt_in: false }], 'Hi')
    assert_equal 0, result[:sent]
  end

  def test_counts_non_eu_active_customers
    result = @sender.send_campaign([{ active: true, region: 'US' }], 'Hi')
    assert_equal 1, result[:sent]
  end

  def test_skips_inactive_customers
    result = @sender.send_campaign([{ active: false, region: 'US' }], 'Hi')
    assert_equal 0, result[:sent]
  end

  def test_handles_mixed_customers
    result = @sender.send_campaign([
      { active: true, region: 'EU', gdpr_opt_in: true },
      { active: true, region: 'EU', gdpr_opt_in: false },
      { active: true, region: 'US' },
      { active: false, region: 'US' }
    ], 'Hi')
    assert_equal 2, result[:sent]
  end

  def test_returns_message_in_result
    result = @sender.send_campaign([], 'Hello World')
    assert_equal 'Hello World', result[:message]
  end

  def test_returns_timestamp_in_result
    result = @sender.send_campaign([], 'Hi')
    assert_kind_of Integer, result[:timestamp]
  end
end

File: csharp/05-zombie-code/test/campaign_sender_test.cs
using Xunit;

public class CampaignSenderTest
{
    [Fact]
    public void CountsEuCustomersWithOptIn()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>
        {
            new CampaignCustomer(true, "EU", true)
        }, "Hi");
        Assert.Equal(1, result.Sent);
    }

    [Fact]
    public void SkipsEuCustomersWithoutOptIn()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>
        {
            new CampaignCustomer(true, "EU", false)
        }, "Hi");
        Assert.Equal(0, result.Sent);
    }

    [Fact]
    public void CountsNonEuActiveCustomers()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>
        {
            new CampaignCustomer(true, "US", false)
        }, "Hi");
        Assert.Equal(1, result.Sent);
    }

    [Fact]
    public void SkipsInactiveCustomers()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>
        {
            new CampaignCustomer(false, "US", false)
        }, "Hi");
        Assert.Equal(0, result.Sent);
    }

    [Fact]
    public void HandlesMixedCustomers()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>
        {
            new CampaignCustomer(true, "EU", true),
            new CampaignCustomer(true, "EU", false),
            new CampaignCustomer(true, "US", false),
            new CampaignCustomer(false, "US", false)
        }, "Hi");
        Assert.Equal(2, result.Sent);
    }

    [Fact]
    public void ReturnsMessageInResult()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>(), "Hello World");
        Assert.Equal("Hello World", result.Message);
    }

    [Fact]
    public void ReturnsTimestampInResult()
    {
        var estimator = new CampaignSender();
        var result = estimator.send_campaign(new List<CampaignCustomer>(), "Hi");
        Assert.True(result.Timestamp > 0);
    }
}
